Design/lines

HaicoDorenbos opened the pull request:
It took quite a long time (3 weeks) before everything was fully fixed, but here is the renovated version of all Line classes which include the following changes:
- All dead code has been marked and out-commented.
- Documentation in all Line classes has been improved.
- All important methods in the Line classes are appropriately tested using equivalence partitioning with boundary value analysis combined with some error guessing.
- Inconsistent methods have been removed, e.g. the length() methods in the Line class, because a Line has infinite length. Also the constructor in the LineSegment class has been removed that accepts a Line, because you have insufficient information to convert a Line to a LineSegment.
- The variable points that encode the Line have been made private and renamed, since in all cases when these points where used it was kind of an anti-pattern where it was more appropriate to use a LineSegment instead.
- The shadow function has been placed at a separate class (I placed this function initially in the LineSegment class which I apologize for).
- A warning message will be sent when creating a Line using the same two points.
- Bugs have been fixed in the projection methods and in the checking whether a point is on the Line.
- Code clones have been removed and methods have been simplified. For this I took running time into account, but cases where the method did not become significantly slower I decided to remove the code clone. However on the other hand cases where it became significantly slower, e.g. intersects versus multiIntersect, I kept the code clone.
- At some places unnecessary copying has been removed, e.g. return Vector2(end - start), since the result of end - start is already a copy.
- The default constructor has been removed, which I consider an anti-pattern, because you should know how a Line, LineSegment, HalfLine looks like before defining it.

Note: Synchronize/utils lines #1141 on the Roboteam_ai branch should be merged, before this PR is merged.

rolfvdhulst commented on include/roboteam_utils/HalfLine.h:
```diff
@@ -18,19 +18,19 @@ class HalfLine {
-         * Construct a new HalfLine object
-         * @param start The location where the HalfLine starts
-         * @param direction This vector indicates in which direction the HalfLine is infinite long
+         * Construct a new HalfLine object by using 2 different points on that Line. Make sure that the given points are different.
```
If you're going to chance the documentation, make it accurate. These points are not just points on the line, but the halline defined by v1,v2 for start direction is different than that which is defined by v2,v1 for start direction.
Also, direction is a confusing name to me, because in the constructor itself you assign it as the position through which the halfline goes. I'd rather then have you describe this constructor as ' construct a new HalfLine starting at start going through direction, which is more accurate. Typically when we talk about a direction of two points v1, v2, I would assume you are talking about e.g. the vector v2-v1. I would also suggest renaming direction to 'goesThrough' or something else than direction.

HaicoDorenbos replied:
I agree to improve the documentation, however for direction I don't have a better name (I rather don't use end as variable name, because it is even more confusing). So if you have a better name tell me.

rolfvdhulst commented on src/Rectangle.cpp:
```diff
@@ -131,6 +131,36 @@ std::vector<Vector2> Rectangle::intersects(const LineSegment &line) const {
+bool Rectangle::contains(const Vector2 &point) const { return maxX() >= point.x && minX() <= point.x && maxY() >= point.y && minY() <= point.y; }
+std::ostream &Rectangle::write(std::ostream &os) const { return os << "Rect: " << corner1 << corner2; }
+
+/* 	 ______   _______  _______  ______     _______  _______  ______   _______
+ *	(  __  \ (  ____ \(  ___  )(  __  \   (  ____ \(  ___  )(  __  \ (  ____ \
+ *	| (  \  )| (    \/| (   ) || (  \  )  | (    \/| (   ) || (  \  )| (    \/
+ *	| |   ) || (__    | (___) || |   ) |  | |      | |   | || |   ) || (__
+ *	| |   | ||  __)   |  ___  || |   | |  | |      | |   | || |   | ||  __)
+ *	| |   ) || (      | (   ) || |   ) |  | |      | |   | || |   ) || (
+ *	| (__/  )| (____/\| )   ( || (__/  )  | (____/\| (___) || (__/  )| (____/\
+ *	(______/ (_______/|/     \|(______/   (_______/(_______)(______/ (_______/
+ *
+ * The functions below are dead. Remove this tag if you use any of the functions and make sure to remove this tag at other places as well that will become alive by using any of the
+ * function below. Do not read/document/redesign/analyse/test/optimize/etc. any of this code, because it is a waste of your time! This code was not removed or placed at another
+ * branch, because other software developers are very attached to this code and are afraid that this code might be used at some day (but I think it won't be used at all and should
+ * be removed).
+ */
```
Is this tag necessary here if it's already in the header?

HaicoDorenbos replied:
Is really necessary, because developers can look only at the cpp file or header file and I should remind them that it is a waste of time to look at this deadcode.

rolfvdhulst commented on src/Shadow.cpp:
```diff
@@ -0,0 +1,39 @@
+        // If both lines from the sources to the start and end of the obstacle do not intersect with the infinite line expansion of this projection line then there is no shadow.
+        return std::nullopt;
+    }
+    std::shared_ptr<LineSegment> shadow = nullptr;
```
Why are you using a shared_ptr?
This makes ownership of this object very weird outside of this function, and LineSegment is small enough that it makes sense to copy it or allow the compiler for e.g. move optimizations.

rolfvdhulst replied:
Just use std::optional if you need the null optionality here.

HaicoDorenbos replied:
You might have a point here, but I will first discuss this in-depth.

rolfvdhulst commented on src/Shadow.cpp:
```diff
@@ -0,0 +1,39 @@
+        Vector2 closest = distanceIntersectionToStart < distanceIntersectionToEnd ? project.start : project.end;
+        Vector2 furthest = distanceIntersectionToStart < distanceIntersectionToEnd ? project.end : project.start;
+        bool furthestInShadow = LineSegment(source, furthest).doesIntersect(obstacle);
+        shadow = std::make_shared<LineSegment>(furthestInShadow ? LineSegment(onlyIntersection, furthest) : LineSegment(onlyIntersection, closest));
```
Noooo

HaicoDorenbos replied:
Yeeeeeesss

rolfvdhulst commented on src/Shadow.cpp:
```diff
@@ -0,0 +1,39 @@
+        bool furthestInShadow = LineSegment(source, furthest).doesIntersect(obstacle);
+        shadow = std::make_shared<LineSegment>(furthestInShadow ? LineSegment(onlyIntersection, furthest) : LineSegment(onlyIntersection, closest));
+    }
+    return shadow->length() > negligible_shadow_length ? std::optional(*shadow) : std::nullopt;
```
Yikes, dereferencing the shared pointer here to make a copy. Just create an object man. Or use std::optional

HaicoDorenbos replied:
No, you should not create an undefined Line object. Only define a Line if you actually know its value...

rolfvdhulst replied:
Ok, so why are you not using an std::optional if you return a nullopt?

std::optional(*shadow) is very very ugly IMO

rolfvdhulst commented on test/LineIntersectionTest.cpp:
```diff
@@ -0,0 +1,237 @@
+#include "roboteam_utils/LineSegment.h"
+#include "roboteam_utils/Vector2.h"
+
+const double EPSILON = 1e-2; // Small difference that is still significant enough to really change a Vector2 position.
```
This is way too high for my liking. Some of my numerical analysis/algorithms teachers recommended me to keep 10^(-12) and 10^(-10) as respective limits for errors. I'd like to see here 10^(-6) here as we might do some floating conversions implicitly somewhere.

HaicoDorenbos replied:
You should note that vectors are similar if they have a Manhattan distance lower than 1e-5. This EPSILON has a quite high value, because you don't want to create similar vector positions.

rolfvdhulst commented on include/roboteam_utils/Line.h:
```diff
@@ -1,101 +1,127 @@
-    public:
+   private:
+    Vector2 location;   // One of the points located on the Line, which represents the location of the Line.
+    Vector2 direction;  // Another point located on the Line, which should be different than the location. It represents the direction of the Line.
```
Again disagree with the naming (see HalfLine). Direction is usually implied to be a difference between two vectors, this is a point on the line. This way if the line is affine and not linear I would totally misuse this class.

HaicoDorenbos replied:
If you have a better naming tell me... (except start and end which is definitely worse)

rolfvdhulst commented on src/Line.cpp:
```diff
@@ -1,33 +1,88 @@
+    this->location = location;
+    this->direction = direction;
+    if (location == direction) {
+        std::cout << "Warning: you created an undefined line, because location == direction. Note that Lines have an infinite length. If you want to have a Line with finite length"
```
use cerr if you are going to throw an error. Red stands out better.

HaicoDorenbos replied:
I agree with you

timovdk replied:
Use the Print utils Lukas made

HaicoDorenbos replied:
Thanks, you prevented me from introducing code clones

rolfvdhulst commented on src/Line.cpp:
```diff
@@ -1,33 +1,88 @@
 
-Vector2 Line::direction() const {
-    return Vector2(end - start);
+float Line::relativePosition(Vector2 line1Location, Vector2 line1Direction, Vector2 pointOnLine) {
+    float xDiff = line1Direction.x - line1Location.x;
+    if (xDiff == 0) {
+        return (pointOnLine.y - line1Location.y) / (line1Direction.y - line1Location.y);
+    } else {
+        return (pointOnLine.x - line1Location.x) / xDiff;
+    }
+}
```
This function is confusing to me, even with the documentation. What are you computing here?

Because if it is the relative position on the line, then there is an error in here, I think, unless you check for parallel and colinear lines somewhere?

HaicoDorenbos replied:
You must assume that pointOnLine is a point on the infinite Line defined by line1Location and line1Direction before using this function. Also you assume that line1Location != line1Direction. What you do now is compute relative position of pointOnLine, i.e. you return k such that pointOnLine = (line1Direction - line1Location) * k + line1Location. I don't think there is any error here, except if you use this function if one of the assumptions does not hold.

HaicoDorenbos replied:
I guess there is no error here. Assuming that you satisfy all preconditions before using this function, i.e.
- line1Location != line1Direction.
- pointOnLine actually lies on the line defined by line1Location and line1Direction.

These preconditions are checked for in other functions that use this function.

rolfvdhulst commented on src/Line.cpp:
```diff
@@ -1,33 +1,88 @@
 
-bool Line::isVertical() const {
-    return (end.x == start.x) && (end.y != start.y);
+std::optional<Vector2> Line::intersect(const Vector2 line1Location, const Vector2 line1Direction, const Vector2 line2Location, const Vector2 line2Direction) {
+    Vector2 A = line1Location - line1Direction;
+    Vector2 B = line2Location - line2Direction;
+    double denom = A.cross(B);
+    if (denom != 0) {
+        Vector2 C = line1Location - line2Location;
+        double numer = C.cross(A);
+        double u = numer / denom;
+        return line2Location - B * u;
+    }
+    return std::nullopt;
 }
```
You're not checking for the case where the two lines overlap or where they are colinear. You always return std::nullopt for colinear lines, which is not true.

HaicoDorenbos replied:
I explained this in the documentation, when two lines are parallel (which includes colinear and overlapping) I don't return any point (because there are either infinitely many intersections or none at all), because of performance reasons. This also happened in the previous implementation.

HaicoDorenbos replied:
That is indeed the behaviour according to the documentation.

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
-std::pair<double, double> LineSegment::coefficients() const {
-    return {slope(), intercept()};
+std::optional<Vector2> LineSegment::intersects(const LineSegment &line) const {
+    auto result = Line::intersect(start, end, line.start, line.end);
```
Oof, since this has a bug now also this code has a bug. Note that here, if your lineSegments are colinear and on the same line the line segments do not necessarily intersect...

rolfvdhulst replied:
This is exactly why I dislike explicit code dependencies and you removing 'duplicate code' . Not only does it make things slower but problems /bugs propagate and create code dependencies which are unnecessary.

HaicoDorenbos replied:
I don't see the problem with code dependencies. They are actually good, because if there is a bug then you immediately notice it because it goes wrong in every function (not just one). Furthermore there is no bug, as mentioned in the documentation intersects return false if the Lines are equal.

rolfvdhulst replied:
Take the case

LineSegment(Vector2(1,1), Vector2(3,3)), LineSegment(Vector2(2,2),Vector2(4,4)) and check it...

HaicoDorenbos replied:
They intersect when I add this test case:
TEST(LineIntersectionTests, Rolf) {
    LineSegment line1(Vector2(1,1), Vector2(3,3));
    LineSegment line2(Vector2(2,2),Vector2(4,4));
    EXPECT_TRUE(line1.intersects(line2));
}

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
+            return std::nullopt;
+        }
+    } else {
+        if (line.isOnLine(start)) {
+            return start;
+        } else if (line.isOnLine(end)) {
+            return end;
+        } else if (isOnLine(line.start)) {
+            return line.start;
```
This type of check also needs to be in the other branch for when the two lines which define the line segments are exactly on the same line (colinear)

HaicoDorenbos replied:
By reasoning you can reason that if start, end and the start point of the other line is not on both lines then also not the end of the other line is on your line.

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
-    return false;
 }
 
+bool LineSegment::operator==(const LineSegment &other) const {
+    return ((start == other.start && end == other.end) || (start == other.end && end == other.start));
+}
```
I dislike this implementation because for Vector2::operator == we have the error bounds. For lines/linesegments if the points are even slightly different this can mean the computations can give vastly different results (e.g. not parallel/parallel). I'd rather have you check for the exact values, personally.

HaicoDorenbos replied:
This function is only used for testing purposes, so probably it is dead code now. I will check if it is actually dead. However I don't agree with you that you have to check for the exact values. Be consistent is the general guideline, so either check for exact values everywhere or don't check for exact values everywhere.

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
+            return std::nullopt;
+        }
+    } else {
+        if (line.isOnLine(start)) {
+            return start;
+        } else if (line.isOnLine(end)) {
+            return end;
+        } else if (isOnLine(line.start)) {
+            return line.start;
```
Personally, I also really dislike the isOnLine() checks because they depend on numerical precision. The previous method had one computation which gave an output which had much less logical steps in it and was actually more consistent than this.

HaicoDorenbos replied:
It was a lot of code cloning, here you have a clear and understandable part of code. But if you know a faster implementation then tell me.

rolfvdhulst commented on test/LineIntersectionTest.cpp:
```diff
@@ -0,0 +1,237 @@
+    Vector2 start1(5.0, 3.0), end1(5.0, 6.0), start2(5.0, 12.0), end2(5.0, 3.0);
+    Line line2(start2, end2); HalfLine halfLine(start1, end1); LineSegment lineSegment1(start1, end1), lineSegment2(start2, end2);
+    std::optional<Vector2> result = halfLine.intersect(line2);
+    EXPECT_FALSE(result.has_value());
+    result = Line::intersect(start1, end1, start2, end2);
+    EXPECT_FALSE(result.has_value());
```
What! They intersect/overlap!!

HaicoDorenbos replied:
Did you read the documentation?

rolfvdhulst replied:
Then I disagree with how you defined intersection/the function. Any user will intuitively expect this is overlapping

HaicoDorenbos replied:
With the non-static intersect variant I would agree to change the definition. But with this one I don't agree, since based on performance it is actually quite fast.

rolfvdhulst commented on test/LineProjectionTest.cpp:
```diff
@@ -0,0 +1,205 @@
+#include "roboteam_utils/LineSegment.h"
+#include "roboteam_utils/Vector2.h"
+
+const double EPSILON = 1e-2; // Small difference that is still significant enough to really change a Vector2 position.
```
Again, please change. This is not an acceptable limit within our code at all.

HaicoDorenbos replied:
It is as explained before.

rolfvdhulst commented on test/LineTest.cpp:
```diff
@@ -10,119 +10,51 @@
-    EXPECT_FALSE(l7.isParallel(l4));
-    EXPECT_FALSE(l7.isParallel(l5));
-    EXPECT_FALSE(l4.isParallel(l8));
+const double EPSILON = 1e-2;  // Small difference that is still significant enough to really change a Vector2 position.
```
Yeah no.

HaicoDorenbos replied:
Yeah yes.

rolfvdhulst commented on test/PolygonTest.cpp:
```diff
@@ -234,6 +240,6 @@ TEST(centroid, PolygonTest) {
     ASSERT_EQ(triangle.centroid(),centroid1);
     ASSERT_EQ(triangle2.centroid(),centroid2);
-    ASSERT_TRUE(quadrilateral.isSimple());
+    // ASSERT_TRUE(quadrilateral.isSimple());
```
Did you remove this just so the function is not used? This is actually a critical part in the testing procedure, so I would probably want this to be readded. If this line fails, the rest of the test might also fail and we would have no idea what is causing it.

HaicoDorenbos replied:
This most likely dead code that is being tested. Never test dead code, because it is a waste of time.

rolfvdhulst commented on test/PolygonTest.cpp:
```diff
@@ -152,8 +158,8 @@ TEST(areas, PolygonTest) {
     Polygon triangleA({A, B, C}), triangleB({A, B, D}), triangleC({A, B, E}), triangleD({A, B, F});
-    ASSERT_TRUE(parallelogram.isSimple());
-    ASSERT_TRUE(rectangle.isSimple());
+    // ASSERT_TRUE(parallelogram.isSimple());
+    // ASSERT_TRUE(rectangle.isSimple());
```
Again, this is CRUCIAL to verify that the test is working properly.

HaicoDorenbos replied:
Never test dead code...

rolfvdhulst replied:
Well it's not dead code. This is literally code that verifies if the test cases are ok.

rolfvdhulst commented on test/ShadowTest.cpp:
```diff
@@ -0,0 +1,38 @@
+#include "roboteam_utils/Vector2.h"
+
+using namespace rtt;
+TEST(LineTests, Shadow) {
+    Vector2 source(1.0, 2.0), projectLine_start(-4.0, -6.0), projectLine_end(-4.0, 6.0), blockLine1_start(0.0, 2.0), blockLine1_end(0.0, 2.0), blockLine2_start(-1.0, 0.0),
+        blockLine2_end(-1.0, 2.0), blockLine3_start(1.0, 1.0), blockLine3_end(0.0, 0.0), blockLine4_start(1.0, 3.0), blockLine4_end(0.0, 2.0), blockLine5_start(0.0, -4.0),
+        blockLine5_end(0.0, 4.0), blockLine6_start(1.0, 2.0), blockLine6_end(1.0, 2.0), blockLine7_start(1.0, 0.0), blockLine7_end(1000000000.0, 0.0), blockLine8_start(-1.0, 1.0),
+        blockLine8_end(-3.0, -1.0), checkLine2_start(-4.0, -3.0), checkLine2_end(-4.0, 2.0), checkLine4_start(-4.0, 2.0), checkLine4_end(-4.0, 6.0), checkLine8_start(-4.0, -0.5),
+        checkLine8_end(-4.0, -1.75);
```
plz make something readable.

HaicoDorenbos replied:
I will think about it...

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
+std::vector<Vector2> LineSegment::multiIntersect(const LineSegment &line) const {
+    auto result = Line::intersect(start, end, line.start, line.end);
+    if (result.has_value()) {
+        float t = Line::relativePosition(start, end, result.value());
+        float u = Line::relativePosition(line.start, line.end, result.value());
```
This is less clear to me than the previous version, to be honest.

HaicoDorenbos replied:
Very clear for me...

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
+        /* The only possible cases are that one/both of the LineSegments are actually points, the LineSegments have a shared LineSegment part or the LineSegments are distinct and
+         * parallel. */
+        std::vector<Vector2> intersections = {};
+        if (line.isOnLine(start)) {
+            intersections.push_back(start);
+        }
+        if (line.isOnLine(end)) {
+            intersections.push_back(end);
+        }
+        if (isOnLine(line.start)) {
+            intersections.push_back(line.start);
+        }
+        if (isOnLine(line.end)) {
+            intersections.push_back(line.end);
         }
```
Big yikes. This is a lot of unnecessary checks that depend on the numerical precision limit. You have already determined that the points are colinear, use this information please.

HaicoDorenbos replied:
I agree in some sense with you, however I should check whether code cloning outweights the (insignificant) performance loss.

rolfvdhulst commented on src/LineSegment.cpp:
```diff
@@ -19,82 +11,113 @@ double LineSegment::length2() const {
+        if (isOnLine(line.end)) {
+            intersections.push_back(line.end);
         }
+        std::sort(intersections.begin(), intersections.end());
```
You're sorting on the absolute length of the starting position, which is probably not intuitive for the user (atleast it is not for me). I would sort on e.g. the distance from start rather than the distance from 0 here.

HaicoDorenbos replied:
Basically this is used to remove duplicate points. I guess I explained in the documentation that this is the order of the results...

rolfvdhulst replied:
std::sort() does not remove duplicate points

HaicoDorenbos replied:
But the function after that removes the points if they are sorted...

rolfvdhulst commented on include/roboteam_utils/LineSegment.h:
```diff
@@ -1,153 +1,162 @@
+     * LineSegments that are actually points. But it is different from the intersects function in the sense that it can return up to 2 intersections. In case:
+     * - The LineSegments do not intersect, it returns an empty list.
+     * - The LineSegments do only intersect once, it returns a list with only that intersection point.
+     * - The LineSegments intersect multiple times (infinitely often), then these intersections form a LineSegment L. In this case it returns the start and end points of
+     *  LineSegment L.
      *
-     * @param line Other line to check against
-     * @return true True if this->slope() == line.slope()
-     * @return false False if this->slope() != line.slope()
+     * @param line The other (finite) LineSegment.
+     * @return Returns a list of the intersections.
      */
-    [[nodiscard]] bool isParallel(const LineSegment &line) const;
+    [[nodiscard]] std::vector<Vector2> multiIntersect(const LineSegment &line) const;
+
```
Why are you seperating the intersection function into two different intersection functions? This seems pretty weird to me, because now the user has the responsibility to check for colinear lines. In the old code this was not the case.

HaicoDorenbos replied:
No, the user does not have the responsibility to check for colinear lines. The only difference is that this function returns multiple intersection points (which is used in the Rectangle class) whereas the other function always return at most 1 intersection point (even if there are more).

HaicoDorenbos commented:
> See comments. I heavily disagree with naming points 'direction' as direction is usually for the vector difference between two points.
> 
> I also disagree quite a lot with how you reimplemented some methods. The IsOnLine() function is used WAY too much. We should compute with as much precision as we can, and then do checks if the intersection points we compute are okay. More dependencies on a numerical precision limit are bad: check at the end of your computation once perhaps.
> 
> There's also some bugs. You often forget to check for the cases where the line segments are colinear in many places.
> 
> To be honest, I would probably advise never to merge major parts of this pull request. The DeadCode tags are fine, but the excessive rewriting of methods introduced bugs and is largely unnecessary. Also, I'd like to point to you that the previous tests were testing for the cases you miss. Also, you're additionally decreasing performance (which is not a major issue, but a consideration) and in many places the code is more sensitive to numerical precision problems. Especially the use of the IsOnLine() function is excessive and should be avoided at all costs if you can do something like computing the relative factor by which a point is on the line.

My goal is to eventually merge this PR, since it improves the code a lot in some points. 

I do agree with you that direction and location are not the best names, but they are definitely better than start and end (since an infinite line doesn't start & end somewhere).

I don't agree with your point that using a function too often is considered a bad practice. It is actually good to use a function in the same class very often. I do however agree with you that if there is a faster implementation with an 'equal' number of code lines then we should take that solution.

I don't agree with you that you should compute with as much precision as you can. You should be consistent overall, so either compute everywhere with as much precision as you can (which causes problems in case of floating rounding errors) or compute everywhere with a lower precision.

HaicoDorenbos commented:
So Rolf has a good point regarding the non-static intersect function of the Line class (the definition of the static intersect won't change however). Therefore I will discuss this with all of you and you can all vote for a definition and the definition with the most votes will be used. What should the intersect function return if two infinite Lines are equal?
1. No intersection point (Pros: the code won't be changed regarding the old specification and this is the fastest method to compute intersections. Cons: it does not satisfy the mathematical definition of intersection).
2. An arbitrary intersection point, like start or end (Pros: quite fast and satisfies mathematical definition. Cons: the function is not symmetric anymore intersect(line1, line2) != intersect(line2, line1) and using an equal Line might return a different results).
3. The intersection point closest to the origin (0, 0) (Pros: satisfies mathematical definition, consistent result regarding symmetry and equal Lines, Cons: quite slow solution).
4. Return a Line class instead of a Vector2 that represents all intersection points (Pros: user can decide which intersection point he actually wants. Cons: user always has to check whether return value is Line or Vector2, slowest definition of them all, take most development time to make code entirely adapt to this new situation).

HaicoDorenbos commented:
I had a discussion with Timo and I concluded that the entire Roboteam_utils repository should be replaced by an external library if possible...

HaicoDorenbos commented:
Everything has been fixed, which includes: 
- Better naming for some variables
- Using optional instead of shared_ptr at the shadow computations
- Better format for error messages
- The intersect functions have been optimized in performance by using the co-linearity assumption
- The new definition for Line intersect has been implemented (i.e. equal Line instances will return the intersection point closest to the origin). 
- Test cases have been updated based on the new definition, also readability of the Shadow test case has been improved on and additional test cases have been created for some functions.
- Dead code that is necessary for testing purposes has been resurrected.
- Documentation has been improved

Additionally I removed (in)equality checks for floats ==, != in these 3 classes and replaced them by an absolute distance check. Furthermore I also made sure that code is re-used at more places.

I haven't removed/fixed: 
- The equality function for LineSegments, since I don't really see the problem with it. 
- The dead code tags at the cpp files, since everywhere it should be mentioned that some code is dead not just in the header file.
- The relativePosition usage in the LineSegment class, because I think it is more readable now and less prone to bugs.
- The epsilon value in the test cases, since you should use a small difference that is still significant different.

HaicoDorenbos commented:
Note: Synchronize/utils lines #1141 on the Roboteam_ai branch should be simultaneously merged.

Yuhanun commented:
$ git fetch origin +refs/pull/76/merge:

First time i ever see travis run this. 

Ignore it for now if it occurs on another merge request then oof
